### PR TITLE
server: fix duplicated records for templates if there are multiple zones

### LINKED ISSUE #8928: Template is not Ready if there are multiple zones
There are two zones in the environment: KVM and VMware

When register a template for all zones, it is stuck at the following status

image

or

image

or

image

##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
API/storage
~~~

##### CLOUDSTACK VERSION

~~~
4.19/4.20
~~~

##### STEPS TO REPRODUCE

~~~

~~~

##### EXPECTED RESULTS

~~~

~~~

##### ACTUAL RESULTS

~~~

~~~

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes #8928

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Review by BryanMLima — APPROVED** [review]
CLGTM, I will try to manually test this today.

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test rocky8 kvm-rocky8

**Review by sureshanaparti — APPROVED** [review]
clgtm

File: server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java [hunk 1]
```diff
@@ -306,12 +306,12 @@ protected void createTemplateWithinZones(TemplateProfile profile, VMTemplateVO t
             zonesIds = _dcDao.listAllZones().stream().map(DataCenterVO::getId).collect(Collectors.toList());
         }
 
-        List<DataStore> imageStores = getImageStoresThrowsExceptionIfNotFound(zonesIds, profile);
 
         for (long zoneId : zonesIds) {
             DataStore imageStore = verifyHeuristicRulesForZone(template, zoneId);
 
             if (imageStore == null) {
+                List<DataStore> imageStores = getImageStoresThrowsExceptionIfNotFound(List.of(zoneId), profile);
```
Comment (BryanMLima) [inline server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java:314]: @weizhouapache, just one remark, with these changes the method `getImageStoresThrowsExceptionIfNotFound` does not need to receive a List, could you refactor to receive only the zone ID?

**Review by BryanMLima — APPROVED** [review]
I manually tested this patch in a local lab, working as intended.

Before two records were created:  
image

After applying the patch, only one record was created:  
image

File: server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java (hunk 1 above)
Comment (weizhouapache) [reply server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java]: @BryanMLima 
yes, thanks. PR is updated

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Pull request merged**